### PR TITLE
Added more Options to Paper

### PR DESCRIPTION
**Pull request by TheShermanTanker**
Changelist is the same as https://github.com/PaperMC/Paper/pull/4067 but this time it's even more polished and every change is in an individual commit

Sorry if I appear like an idiot this is pretty much the first time I've tried a Pull Request

### CODE REVIEW

**Comment by Proximyst** [comment]
At the very least compile your changes before submitting the PR.

Also, please split these changes into separate PRs for easing the merger's life in selecting which are accepted and which aren't. Join our Discord if you'd like more help, as there are still several issues you need to fix.

**Comment by TheShermanTanker** [comment]
> 
> 
> At the very least compile your changes before submitting the PR.
> 
> Also, please split these changes into separate PRs for easing the merger's life in selecting which are accepted and which aren't. Join our Discord if you'd like more help, as there are still several issues you need to fix.

I missed out one line and everything broke :/  

As for the PRs I was kind of afraid of spamming several threads into the PR section, I'll try doing that if I have time

I am in the Paper Discord, by the way

**Comment by kennytv** [comment]
Readding legacy gamemode ids has nothing to do in a Paper patch (and your spacing is still very much bork)🥶

**Comment by MrIvanPlays** [comment]
When seeing someone use spaces instead of tabs :eyes:

**Comment by Draycia** [comment]
Ivan what are you on about? Any sane person uses spaces instead of tabs.
Look in any patch in your fork and tell me if you find a single tab character.

**Comment by MrIvanPlays** [comment]
> Look in any patch in your fork and tell me if you find a single tab character.

image
bruh

**Comment by Draycia** [comment]
image
Looks like a space to me.

**Comment by JanTuck** [comment]
Ivan IJ defaults to converting tabs to 5 or so spaces.

File: Spigot-API-Patches/0219-Add-Operator-API.patch [hunk 1]
```diff
@@ -0,0 +1,29 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: BuildTools <unconfigured@null.spigotmc.org>
```
Comment (CDFN) [inline Spigot-API-Patches/0219-Add-Operator-API.patch:2]: Wrong author

File: Spigot-Server-Patches/0551-Updated-Paper-Config.patch [hunk 2]
```diff
@@ -0,0 +1,41 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: BuildTools <unconfigured@null.spigotmc.org>
```
Comment (CDFN) [inline Spigot-Server-Patches/0551-Updated-Paper-Config.patch:2]: Wrong author

File: Spigot-Server-Patches/0552-Removed-redundant-Patrol-Option-and-Fixed-Patrol-Spa.patch [hunk 3]
```diff
@@ -0,0 +1,23 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: BuildTools <unconfigured@null.spigotmc.org>
```
Comment (CDFN) [inline Spigot-Server-Patches/0552-Removed-redundant-Patrol-Option-and-Fixed-Patrol-Spa.patch:2]: Wrong author

File: Spigot-Server-Patches/0553-Trader-Spawning-now-depends-on-specific-gamerule.patch [hunk 4]
```diff
@@ -0,0 +1,30 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: BuildTools <unconfigured@null.spigotmc.org>
```
Comment (CDFN) [inline Spigot-Server-Patches/0553-Trader-Spawning-now-depends-on-specific-gamerule.patch:2]: Wrong author

File: Spigot-Server-Patches/0553-Trader-Spawning-now-depends-on-specific-gamerule.patch [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
++++ b/src/main/java/net/minecraft/server/MobSpawnerTrader.java
+@@ -41,11 +41,7 @@ public class MobSpawnerTrader implements MobSpawner {
+                 return 0;
+             } else {
+                 this.d = 24000;
+-                if (!worldserver.getGameRules().getBoolean(GameRules.DO_MOB_SPAWNING)) {
+-                    return 0;
+-                } else {
+                     int i = this.e;
+-
```
Comment (CDFN) [inline Spigot-Server-Patches/0553-Trader-Spawning-now-depends-on-specific-gamerule.patch:19]: No need to remove this line (keep minimal diff as it is only possible0

File: Spigot-Server-Patches/0556-Operator-API-Impl.patch [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
+index 9382e8f79e8edec8885c629a36e230fbec50e1fb..fda45f3e4588c46d40ccedf5ef1fd1dbc2b70ae3 100644
+--- a/src/main/java/net/minecraft/server/PlayerList.java
++++ b/src/main/java/net/minecraft/server/PlayerList.java
+@@ -1011,6 +1011,19 @@ public abstract class PlayerList {
+         return this.l;
+     }
+ 
++    public void setOPLevel(GameProfile profile, int level) {
```
Comment (CDFN) [inline Spigot-Server-Patches/0556-Operator-API-Impl.patch:15]: Keep convention - in method below you have add**Op**, here you have set**OP**Level.

**Comment by MrIvanPlays** [comment]
> Ivan IJ defaults to converting tabs to 5 or so spaces.

Looks like I need to change settings

File: Spigot-Server-Patches/0556-Operator-API-Impl.patch [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+index 5770d4183c1b9ab6119a25930283c0235250ed6e..631eb25ef344818e349b401133ecc7f8da61efd5 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+@@ -525,4 +525,14 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
+             manager.save();
+         }
+     }
++
++	@Override
++	public int getOPLevel() {
```
Comment (CDFN) [inline Spigot-Server-Patches/0556-Operator-API-Impl.patch:41]: OP -> Op

File: Spigot-Server-Patches/0556-Operator-API-Impl.patch [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
++		return server.getServer().b(getProfile());
++	}
++
++	@Override
++	public void setOPLevel(int level) {
```
Comment (CDFN) [inline Spigot-Server-Patches/0556-Operator-API-Impl.patch:46]: OP -> Op

File: Spigot-Server-Patches/0556-Operator-API-Impl.patch [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+index 17aa866252a95a9d9a22f0b0b98f2f9d2c700bb5..2925415133af82434fbb82efbc80486cc815a93d 100644
+--- a/src/main/java/org/bukkit/craftbukkit/command/CraftBlockCommandSender.java
++++ b/src/main/java/org/bukkit/craftbukkit/command/CraftBlockCommandSender.java
+@@ -59,4 +59,14 @@ public class CraftBlockCommandSender extends ServerCommandSender implements Bloc
+     public CommandListenerWrapper getWrapper() {
+         return block;
+     }
++
++	@Override
++	public int getOPLevel() {
```
Comment (CDFN) [inline Spigot-Server-Patches/0556-Operator-API-Impl.patch:60]: OP -> Op

File: Spigot-Server-Patches/0556-Operator-API-Impl.patch [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
++		return 4;
++	}
++
++	@Override
++	public void setOPLevel(int level) {
```
Comment (CDFN) [inline Spigot-Server-Patches/0556-Operator-API-Impl.patch:65]: OP -> Op

Comment (CDFN) [reply Spigot-Server-Patches/0556-Operator-API-Impl.patch:65]: Further you have more OPs, won't mark it but change it too.

**Review by CDFN — COMMENTED** [review]
I'm sure 0551 won't get merged - it shouldn't be new patch for changing config. One feature = one commit - it'll allow you to keep patches clean. 
If I'm wrong in any review please correct me =)

**Pull request closed without merging**